include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -3409,8 +3400,114 @@ FEEvaluationBase<dim, n_components_, Number, is_face>::read_cell_data(
+  template <typename T>
+  struct has_local_element
+  {
+  private:
+    // this will work always.
+    // we let it be void as we know T::local_element() (if exists) should
+    // certainly return something
+    static void
+    detect(...);
+
+    // this detecter will work only if we have "... T::local_element() const"
+    // and its return type will be the same as local_element(),
+    // that we expect to be T::value_type
+    template <typename U>
+    static decltype(std::declval<U const>().local_element(0))
+    detect(const U &);
+
+  public:
+    // finally here we check if our detector has return type same as
+    // T::value_type. This will happen if compiler can use second detector,
+    // otherwise SFINAE let it work with the more general first one that is void
+    static constexpr bool value =
+      std::is_same<typename T::value_type,
+                   decltype(detect(std::declval<T>()))>::value;
+  };
+
+
+  // same as above to check
+  // bool T::partitioners_are_compatible(const Utilities::MPI::Partitioner &)
+  // const
+  template <typename T>
+  struct has_partitioners_are_compatible
+  {
+  private:
+    static void
+    detect(...);
+
+    template <typename U>
+    static decltype(std::declval<U const>().partitioners_are_compatible(
+      std::declval<Utilities::MPI::Partitioner>()))
+    detect(const U &);
+
+  public:
+    static constexpr bool value =
+      std::is_same<bool, decltype(detect(std::declval<T>()))>::value;
+  };
+
+
+
+  // same as above to check
+  // T::const_iterator T::begin() const
+  template <typename T>
+  struct has_begin
+  {
+  private:
+    static void
+    detect(...);
+
+    template <typename U>
+    static decltype(std::declval<U const>().begin())
+    detect(const U &);
+
+  public:
+    static constexpr bool value =
+      std::is_same<typename T::const_iterator,
+                   decltype(detect(std::declval<T>()))>::value;
+  };
+
+
+
+  // type trait for vector T and Number to see if
+  // we can do vectorized load/save.
+  // for VectorReader and VectorDistributorLocalToGlobal we assume that
+  // if both begin() and local_element()
+  // exist, then begin() + offset == local_element(offset)
+  template <typename T, typename Number>
+  struct is_vectorizable
+  {
+    static constexpr bool value =
+      has_begin<T>::value &&
+      (has_local_element<T>::value || is_serial_vector<T>::value) &&
+      std::is_same<typename T::value_type, Number>::value;
+  };
+
+
+
+  // access to generic const vectors that have operator ().
+  // FIXME: this is wrong for Trilinos/Petsc MPI vectors
+  // where we should first do Partitioner::local_to_global()
+  template <typename VectorType,
+            typename std::enable_if<!has_local_element<VectorType>::value,
+                                    VectorType>::type * = nullptr>
+  inline typename VectorType::value_type
+  vector_access(const VectorType &vec, const unsigned int entry)
+  {
+    return vec(entry);
```

Comment:
The question is how we can introduce the local-to-global interface. We probably should pass the partitioner from DoFInfo to this call. Can you open an issue for that so we don't forget it?

Reply:
Done https://github.com/dealii/dealii/issues/7753